Move to Java 17

snjeza opened the pull request:
Fixes #2117 

You can test the PR locally.
Jenkins fails because it uses Java 11

~~SendNotificationTest tests~~ and AdvancedOrganizeImportsHandlerTest.testAmbiguousStaticImports require review.

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

rgrunber commented:
Nice. I agree with most of this. What about the `Bundle-RequiredExecutionEnvironment: JavaSE-11` statements in our manifests? Should we not change them to `JavaSE-17` given we can't really run on a lesser version after this ?

snjeza commented:
> Should we not change them to JavaSE-17 given we can't really run on a lesser version after this ?

I have changed it.

rgrunber commented:
There's a few `is missing the Java nature` failures.

[mvn-build-of-pr.txt](https://github.com/eclipse/eclipse.jdt.ls/files/8835856/log.txt)

snjeza commented:
> There's a few is missing the Java nature failures.

Checking...

rgrunber commented:
Most of the errors seem to be resolve and I was able to run once with everything passing. However, on some occasions I see the following failures/errors :

```
ProjectUtilsTest.testGetMaxProjectProblemSeverity:51 expected:<0> but was:<1>

GradleProjectImporterTest.importGradleKtsProject:509->AbstractProjectsManagerBasedTest.assertNoErrors:309 kradle has errors:
Type=org.eclipse.jdt.core.problem:Message=The project was not built since its build path is incomplete. Cannot find the class file for java.lang.Object. Fix the build path then try building this project:LineNumber=null, Type=org.eclipse.jdt.core.problem:Message=The type java.lang.Object cannot be resolved. It is indirectly referenced from required .class files:LineNumber=1 expected:<0> but was:<2>
  MavenProjectImporterTest.testJava17Project:313->AbstractProjectsManagerBasedTest.assertHasErrors:299 Preview features enabled at an invalid source release level was not found in:
Type=org.eclipse.jdt.core.problem:Message=The project was not built due to "Failed to init ct.sym for /home/rgrunber/git/eclipse.jdt.ls/org.eclipse.jdt.ls.tests/fakejdk2/17a/lib/jrt-fs.jar". Fix the problem, then try refreshing this project and building it since it may be inconsistent:LineNumber=null

org.eclipse.jdt.ls.core.internal.handlers.DocumentLifeCycleHandlerTest.testReconcile(org.eclipse.jdt.ls.core.internal.handlers.DocumentLifeCycleHandlerTest)
  Run 1: DocumentLifeCycleHandlerTest.setup:101->mockPreferences:198 WrongTypeOfReturnValue
  Run 2: DocumentLifeCycleHandlerTest.tearDown:114 NullPointer Cannot invoke "org.eclip...
```

testforstephen commented on org.eclipse.jdt.ls.tests.syntaxserver/META-INF/MANIFEST.MF:
```diff
@@ -3,10 +3,11 @@ Bundle-ManifestVersion: 2
 Import-Package: org.osgi.framework;version="1.3.0"
 Bundle-Localization: plugin
 Bundle-ActivationPolicy: lazy
+Import-Package: org.osgi.framework;version="1.3.0"
```
Line 10 duplicates with Line 7.

testforstephen commented:
We also need to update the minimum JDK requirement in client side in case someone is using `java.jdt.ls.java.home` to specify tooling JDK.

https://github.com/redhat-developer/vscode-java/blob/6ec891f313d2ee30f754c80914e9a7e618b942d9/src/requirements.ts#L12

snjeza commented:
>Most of the errors seem to be resolve and I was able to run once with everything passing. However, on some occasions I see the following failures/errors :

@rgrunber I can reproduce it. Working on that.

snjeza commented:
test this please

snjeza commented:
test this please

snjeza commented:
> We also need to update the minimum JDK requirement in client side in case someone is using java.jdt.ls.java.home to specify tooling JDK.

@testforstephen I have created https://github.com/redhat-developer/vscode-java/pull/2495

snjeza commented:
> Most of the errors seem to be resolve and I was able to run once with everything passing. However, on some occasions I see the following failures/errors :

@rgrunber I have updated the PR. Could you, please, test again? I have also updated [Jenkins job](https://ci.eclipse.org/ls/job/jdt-ls-pr) to JDK 17.

snjeza commented:
test this please

rgrunber commented:
Similar to https://ci.eclipse.org/ls/job/jdt-ls-pr/3322/testReport/junit/org.eclipse.jdt.ls.core.internal.handlers/CompletionHandlerTest/testCompletion_AnonymousTypeMoreMethods/ , I saw :

```
Failures: 
  CompletionHandlerTest.testCompletion_AnonymousType:2147 expected:<IFoo() {
[	${0}]
};> but was:<IFoo() {
[]
};>
```

locally. Everything else is passing.

snjeza commented:
> Similar to https://ci.eclipse.org/ls/job/jdt-ls-pr/3322/testReport/junit/org.eclipse.jdt.ls.core.internal.handlers/CompletionHandlerTest/testCompletion_AnonymousTypeMoreMethods/ , I saw :

CompletionHandlerTest fails randomly 
- https://ci.eclipse.org/ls/job/jdt-ls-pr/3321/ (failed)
- https://ci.eclipse.org/ls/job/jdt-ls-pr/3322/ (failed)
- https://ci.eclipse.org/ls/job/jdt-ls-pr/3323 (passed)

I will check it.

snjeza commented:
test this please

rgrunber commented:
I think this is in a good state now, and ready to be merged. I would just ask for @testforstephen to confirm.

@snjeza , I would just add the following to the message body (below the title) :
```
- Update to Tycho 2.7.2
- Update to jacoco-maven-plugin 0.8.7
- Update to org.mockito.mockito-core 4.x
- Update BREE from JavaSE-11 to JavaSE-17
- Use Gradle 7.3 for test projects to ensure compatibility with Java 17
```

We may want to later revisit if it's possible to specify a different JDK than the one used to launch Maven (or Tycho Surefire) for the Gradle daemon used for a particular test. This way we could optionally continue to test older versions of Gradle. Could we have just used the "gradle java home" property through preferences ?

For now I think we should get this change in to get the builds working again.

snjeza commented:
> ...  I would just add the following to the message body (below the title) :

@rgrunber I have added it

snjeza commented:
> ...Any additional modifications we can make once the build is back to normal.

@rgrunber I have updated the [jdt-ls-master](https://ci.eclipse.org/ls/job/jdt-ls-master) build.

afriza commented:
> Nice. I agree with most of this. What about the `Bundle-RequiredExecutionEnvironment: JavaSE-11` statements in our manifests? Should we not change them to `JavaSE-17` given we can't really run on a lesser version after this ?

Which last extension version supports < `JavaSE-17` ?
Because of extension auto-update, I get the install JDK 17 prompt.
